designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+
+Each user-defined constraint within the constraint block must specify a set of functions that define the transformation between unconstrained and constrained spaces, validate data, and increment the log of the absolute value of the determinant of the Jacobian. The general structure of a user-defined constraint is as follows:
+
+```stan
+constraints {
+ constraint_name {
+   // Transform from unconstrained to constrained space
+   ReturnType constrain(TransformType, OtherArgs...) {
+     // ... transformation logic ...
+     // Increment Jacobian
+     jacobian += ...;
+     return ...;
+   }
+
+
+   // Transform from constrained to unconstrained space
+   type unconstrain(TransformType, OtherArgs...) {
+     // ... inverse transformation logic ...
+     return ...;
+   }
+
+
+   // Validate data
+   int validate(TransformType, OtherArgs...) {
+     // ... validation logic ...
+     return ...; // typically returns a boolean value
+   }
+ }
+}
+```
+
+Each constraint type must define a set of functions that govern the transformation and validation of variables subject to the constraint. The first argument for all functions defined in a constraint must be the argument of interest to be transformed or validated. These functions include:
+
+- `constrain(TransformType, OtherArguments...) -> ReturnType`: Transforms a real unconstrained variable to a constrained space.
+- `unconstrain(TransformType, OtherArguments...) -> ReturnType`: Transforms a real constrained variable back to the unconstrained space.
+- `validate(TransformType, OtherArguments...) -> ReturnType`: Validates a real variable against the constraint, returning 1 if valid and 0 otherwise.
+
+### Using the jacobian Keyword
+
+The `jacobian` keyword is introduced within the Constraint block to facilitate the increment of the log of the absolute determinant of the Jacobian. This keyword behaves similarly to the `target` keyword, allowing users to account for the change of variables in the probability density function. The `jacobian` keyword is available in both the `constraints` block and the `transformed parameters` block.
+
+### Declaring Variables with User-defined Constraints
+
+Variables can be declared with user-defined constraints in the data, transformed data, parameters, transformed parameters, and generated quantities blocks. The syntax for declaring a variable with a user-defined constraint is as follows:
+
+```stan
+type<udc=constraint_name(args...)> variable_name;
+```
+
+Here, udc denotes a user-defined constraint , and constraint_name(args...) specifies the name of the constraint along with any arguments it requires.
+
+### Example Usage
+
+```stan
+constraints {
+ upper_bound {
+   real constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real constrain(vector x, real upper_bound) {
```

Comment:
Return should be `vector` here.

Now it looks like you're going to define constrained containers directly rather than mapping?

Reply:
👍 

> Now it looks like you're going to define constrained containers directly rather than mapping?

Sorry can you clarify? Not sure what you mean